bin/ocamlformat-rpc/main.ml

```diff
@@ -0,0 +1,127 @@
+(*                              OCamlFormat                               *)
+(*                                                                        *)
+(*            Copyright (c) Facebook, Inc. and its affiliates.            *)
+(*                                                                        *)
+(*      This source code is licensed under the MIT license found in       *)
+(*      the LICENSE file in the root directory of this source tree.       *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** OCamlFormat-RPC *)
+
+open Ocamlformat_lib
+
+;;
+Caml.at_exit (Format.pp_print_flush Format.err_formatter)
+
+;;
+Caml.at_exit (Format_.pp_print_flush Format_.err_formatter)
+
+module V = struct
+  type t = V1
+
+  let is_handled = function "v1" | "V1" -> Some V1 | _ -> None
+
+  let propose_another = function "v1" | "V1" -> None | _ -> Some V1
+
+  let to_string = function V1 -> "v1"
+end
+
+type state = Waiting_for_version | Version_defined of V.t
+
+let format fg source =
+  let input_name = "<rpc input>" in
+  let conf = Conf.default_profile in
+  let opts =
+    Conf.{debug= false; margin_check= false; format_invalid_files= false}
+  in
+  Translation_unit.parse_and_format fg ~input_name ~source conf opts
+
+let rec rpc_main = function
+  | Waiting_for_version -> (
+    match Ocamlformat_rpc_lib.Init.read_input stdin with
+    | `Halt -> Ok ()
+    | `Unknown -> Ok ()
+    | `Version vstr -> (
+      match V.is_handled vstr with
+      | Some v ->
+          Ocamlformat_rpc_lib.Init.output stdout (`Version vstr) ;
+          Out_channel.flush stdout ;
+          rpc_main (Version_defined v)
+      | None -> (
+        match V.propose_another vstr with
+        | Some v ->
+            let vstr = V.to_string v in
+            Ocamlformat_rpc_lib.Init.output stdout (`Version vstr) ;
+            rpc_main Waiting_for_version
+        | None -> Ok () ) ) )
+  | Version_defined v as state -> (
+    match v with
+    | V1 -> (
+      match Ocamlformat_rpc_lib.V1.read_input stdin with
+      | `Halt -> Ok ()
+      | `Unknown | `Error _ -> rpc_main state
+      | `Format x ->
+          List.fold_until ~init:()
+            ~finish:(fun () ->
+              Ocamlformat_rpc_lib.V1.output stdout
+                (`Error (Format.flush_str_formatter ())) )
+            ~f:(fun () try_formatting ->
+              match try_formatting () with
+              | Ok formatted ->
+                  ignore (Format.flush_str_formatter ()) ;
+                  Ocamlformat_rpc_lib.V1.output stdout (`Format formatted) ;
+                  Stop ()
+              | Error e ->
+                  Translation_unit.Error.print ~fmt:Format.str_formatter
+                    ~exe:"ocamlformat-rpc" ~debug:false ~quiet:false
+                    ~input_name:x e ;
+                  Continue () )
+            [ (fun () -> format Signature x)
```

Comment:
Note that some input could format successfully with multiple "kind" and that we have no guarantee that they would format the same way.

Reply:
Yes, I tried to order the formatting functions in such a way that the ones expecting a keyword first (like signatures) are placed before the more general ones (like toplevel phrases), but I try to find ambiguous cases and add them in the tests.

Reply:
Aren't you missing Structure here ?

Reply:
It won't be used when called by ocaml-lsp, that should use types, module types and signatures, and mdx should use only toplevel phrases, and maybe signatures.
And parsing a file as `--impl` in ocamlformat processes it as a use_file (toplevel phrases) anyway.

Reply:
Adding a small comment about this will certainly help future contributers

Reply:
done!